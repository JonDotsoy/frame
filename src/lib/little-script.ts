import { inspect } from "util"

export namespace LittleScript {
  export abstract class Sentence {
    abstract toSource(): string
  }
  export class LiteralSentence extends Sentence {
    constructor(readonly source: string) {
      super()
    }
    toSource(): string {
      return this.source
    }
  }
  export class AwaitSentence extends Sentence {
    constructor(readonly sentence: CallSentence) {
      super()
    }
    toSource(): string {
      return `await ${this.sentence.toSource()}`
    }
  }
  export class AsyncFunctionSentence extends Sentence {
    constructor(
      readonly functionName: IdentifierSentence,
      readonly args: any[],
      readonly body: ProgramSentence
    ) {
      super()
    }
    toSource(): string {
      return [
        `async function ${this.functionName.toSource()}() {`,
        this.body.toSource({ ident: 2 }),
        "}",
      ].join("\n")
    }
  }
  export class ImportSentence extends Sentence {
    constructor(
      readonly asIdentifier: IdentifierSentence,
      readonly source: string
    ) {
      super()
    }
    toSource(): string {
      return `import * as ${this.asIdentifier.toSource()} from ${inspect(
        this.source
      )}`
    }
  }
  export class ExportSentence extends Sentence {
    constructor(readonly sentence: VariableSentence) {
      super()
    }
    toSource(): string {
      return `export ${this.sentence.toSource()}`
    }
  }
  export class IdentifierSentence extends Sentence {
    constructor(readonly literal: string) {
      super()
    }
    toSource(): string {
      return this.literal
    }
  }
  export class VariableSentence extends Sentence {
    constructor(
      readonly type: "let" | "const",
      readonly variableName: IdentifierSentence,
      readonly assign:
        | AwaitSentence
        | IdentifierSentence
        | MapSentence
        | CallSentence
    ) {
      super()
    }
    toSource(): string {
      return `${
        this.type
      } ${this.variableName.toSource()} = ${this.assign.toSource()}`
    }
  }
  export class MapSentence extends Sentence {
    constructor(
      readonly k: IdentifierSentence,
      readonly v: IdentifierSentence
    ) {
      super()
    }
    toSource(): string {
      return `new Map<${this.k.toSource()}, ${this.v.toSource()}>()`
    }
  }
  export class GetParamSentence extends Sentence {
    constructor(
      readonly parent: IdentifierSentence,
      readonly prop: IdentifierSentence
    ) {
      super()
    }
    toSource(): string {
      return `${this.parent.toSource()}.${this.prop.toSource()}`
    }
  }
  export class ArraySentence extends Sentence {
    constructor(readonly params: (IdentifierSentence | GetParamSentence)[]) {
      super()
    }
    toSource(): string {
      return `[${this.params.map((param) => param.toSource()).join(", ")}]`
    }
  }
  export class CallSentence extends Sentence {
    constructor(
      readonly parent: null | IdentifierSentence | CallSentence,
      readonly method: IdentifierSentence,
      readonly params: (IdentifierSentence | GetParamSentence | ArraySentence)[]
    ) {
      super()
    }
    toSource(): string {
      const p = this.parent
        ? `${this.parent.toSource()}.${this.method.toSource()}`
        : this.method.toSource()

      return `${p}(${this.params.map((param) => param.toSource()).join(", ")})`
    }
  }
  type ProgramSentenceTypesAllow =
    | AwaitSentence
    | AsyncFunctionSentence
    | LiteralSentence
    | VariableSentence
    | ExportSentence
    | ImportSentence
    | MapSentence
    | CallSentence
  export class ProgramSentence extends Sentence {
    constructor(readonly body: ProgramSentenceTypesAllow[] = []) {
      super()
    }
    append(sentence: ProgramSentenceTypesAllow) {
      this.body.push(sentence)
    }
    toSource(opts?: { ident?: number }): string {
      const ident = opts?.ident ?? 0
      const identStr = " ".repeat(ident)
      return this.body.map((item) => `${identStr}${item.toSource()}`).join("\n")
    }
  }
}
