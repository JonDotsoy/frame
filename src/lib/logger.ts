import { debuglog } from "util"

export class Logger {
  private nodeLog: import("util").DebugLogger

  constructor(namespace: string) {
    this.nodeLog = debuglog(namespace)
  }

  log(message: string) {
    this.nodeLog(message)
  }
}
