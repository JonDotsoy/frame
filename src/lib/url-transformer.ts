import { relative } from "path"
import { fileURLToPath } from "node:url"

export class URLTransformer {
  constructor(readonly url: URL, readonly urlSource?: URL | string) { }

  toURL() {
    return this.url
  }

  toDirname() {
    return new URLTransformer(new URL('./', this.url))
  }

  toPath() {
    return fileURLToPath(this.toURL())
  }

  toString(): string {
    return this.url.toString()
  }

  toRelativeFrom(from: string): string {
    return relative(from, this.url.pathname)
  }

  static fromPlainURL(urlLike: string): URLTransformer {
    return new URLTransformer(new URL(urlLike))
  }
}
