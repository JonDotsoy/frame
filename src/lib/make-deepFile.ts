import { basename, dirname, relative } from "path"
import { writeFile } from "fs/promises"
import { Cache } from "./dto/cache"
import { LittleScript } from "./little-script"

export interface MakeDeepFileOptions {
  makeBootstrap?: boolean
}

export const makeDeepFile = async (
  outLocation: string,
  cache: Cache,
  options?: MakeDeepFileOptions
) => {
  const makeBootstrap = options?.makeBootstrap ?? false
  const program = new LittleScript.ProgramSentence()
  const deepsIdentifier = new LittleScript.IdentifierSentence("deeps")
  const setIdentifier = new LittleScript.IdentifierSentence("set")

  const importIdentifierSourceLocations = new Map<
    string,
    LittleScript.IdentifierSentence
  >()
  const importIdentifierNames = new Set<string>(["deeps", "ctx", "bootstrap"])

  const toImportIdentifier = (
    sourceLocation: string
  ): LittleScript.IdentifierSentence => {
    const importIdentifier = importIdentifierSourceLocations.get(sourceLocation)
    if (!importIdentifier)
      throw new Error(`Cannot found identifier to ${sourceLocation}`)
    return importIdentifier
  }

  const toRelativeSource = (sourceLocation: string) => {
    const _sourceRelativeLocation = relative(
      dirname(outLocation),
      sourceLocation
    )
    const __sourceRelativeLocation =
      _sourceRelativeLocation.startsWith("/") ||
      _sourceRelativeLocation.startsWith(".")
        ? _sourceRelativeLocation
        : `./${_sourceRelativeLocation}`
    const sourceRelativeLocation = __sourceRelativeLocation.endsWith(".ts")
      ? __sourceRelativeLocation.substring(
          0,
          __sourceRelativeLocation.length - 3
        )
      : __sourceRelativeLocation

    const sourceBasename = basename(sourceRelativeLocation).replace(/\W/g, "_")

    const eachName = (name: string): string => {
      let iter = 0
      let keyword = name

      while (importIdentifierNames.has(keyword)) {
        iter = iter + 1
        keyword = `${name}_${iter}`
      }

      return keyword
    }

    const keyword = eachName(sourceBasename)

    return { keyword, sourceRelativeLocation }
  }

  const appendResourceImport = (sourceLocation: string) => {
    const relativeSource = toRelativeSource(sourceLocation)
    const scriptIdentified = new LittleScript.IdentifierSentence(
      relativeSource.keyword
    )
    importIdentifierNames.add(relativeSource.keyword)
    importIdentifierSourceLocations.set(sourceLocation, scriptIdentified)

    program.append(
      new LittleScript.ImportSentence(
        scriptIdentified,
        relativeSource.sourceRelativeLocation
      )
    )

    return {
      relativeSource,
      scriptIdentified,
    }
  }

  // load tom
  const tomSourceLocation = require.resolve("../tom")
  const tomResourceImport = appendResourceImport(tomSourceLocation)
  const tomRelativeSource = tomResourceImport.relativeSource
  const tomScriptIdentified = tomResourceImport.scriptIdentified

  for (const [sourceLocation, module] of cache) {
    appendResourceImport(sourceLocation)
  }

  program.append(new LittleScript.LiteralSentence(""))

  program.append(
    new LittleScript.VariableSentence(
      "const",
      deepsIdentifier,
      new LittleScript.MapSentence(
        new LittleScript.IdentifierSentence("any"),
        new LittleScript.IdentifierSentence("any[]")
      )
    )
  )

  program.append(new LittleScript.LiteralSentence(""))

  for (const [sourceLocation, module] of cache) {
    const importIdentifier = toImportIdentifier(sourceLocation)

    for (const [className, { parameters }] of Object.entries(module.classes)) {
      program.append(
        new LittleScript.CallSentence(deepsIdentifier, setIdentifier, [
          new LittleScript.GetParamSentence(
            importIdentifier,
            new LittleScript.IdentifierSentence(className)
          ),
          new LittleScript.ArraySentence(
            parameters.map((parameter) => {
              return new LittleScript.GetParamSentence(
                toImportIdentifier(parameter.module),
                new LittleScript.IdentifierSentence(parameter.exportName)
              )
            })
          ),
        ])
      )
    }
  }

  program.append(new LittleScript.LiteralSentence(""))

  const ctxIdentifier = new LittleScript.IdentifierSentence("ctx")
  program.append(
    new LittleScript.ExportSentence(
      new LittleScript.VariableSentence(
        "const",
        ctxIdentifier,
        new LittleScript.CallSentence(
          tomScriptIdentified,
          new LittleScript.IdentifierSentence("createContext"),
          [deepsIdentifier]
        )
      )
    )
  )

  program.append(new LittleScript.LiteralSentence(""))

  if (makeBootstrap) {
    const inferClassApp = ():
      | { location: string; className: string }
      | undefined => {
      for (const [location, cacheItem] of cache.entries()) {
        for (const [classRef, ref] of Object.entries(cacheItem.classes)) {
          if (classRef.toLowerCase() === "app") {
            return {
              location,
              className: classRef,
            }
          }
        }
      }
    }

    const c = inferClassApp()
    if (!c) throw new Error("Not posible infer the class App")

    const bootstrapVariableIdentifier = new LittleScript.IdentifierSentence(
      "bootstrap"
    )
    const myAppVariableIdentifier = new LittleScript.IdentifierSentence("myApp")
    const newLocal_1 = new LittleScript.VariableSentence(
      "const",
      myAppVariableIdentifier,
      new LittleScript.AwaitSentence(
        new LittleScript.CallSentence(
          ctxIdentifier,
          new LittleScript.IdentifierSentence("get"),
          [
            new LittleScript.GetParamSentence(
              toImportIdentifier(c.location),
              new LittleScript.IdentifierSentence(c.className)
            ),
          ]
        )
      )
    )

    program.append(
      new LittleScript.AsyncFunctionSentence(
        bootstrapVariableIdentifier,
        [],
        new LittleScript.ProgramSentence([
          newLocal_1,
          new LittleScript.AwaitSentence(
            new LittleScript.CallSentence(
              myAppVariableIdentifier,
              new LittleScript.IdentifierSentence("run"),
              []
            )
          ),
        ])
      )
    )

    program.append(new LittleScript.LiteralSentence(""))

    program.append(
      new LittleScript.CallSentence(
        new LittleScript.CallSentence(null, bootstrapVariableIdentifier, []),
        new LittleScript.IdentifierSentence("catch"),
        [
          new LittleScript.GetParamSentence(
            new LittleScript.IdentifierSentence("console"),
            new LittleScript.IdentifierSentence("error")
          ),
        ]
      )
    )

    program.append(new LittleScript.LiteralSentence(""))
  }

  await writeFile(outLocation, program.toSource(), "utf-8")
}
