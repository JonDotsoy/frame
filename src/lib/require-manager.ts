import { existsSync, PathLike } from "fs"
import { Logger } from "./logger"
import { URLTransformer } from "./url-transformer"

type ExtDef = Record<string, { rewriteExt: string }>

const logger = new Logger("tomts")

export class RequireManager {
  readonly exts: ExtDef = {}

  constructor(readonly baseLocation: URL, options?: { exts?: ExtDef }) {
    this.exts = {
      "": { rewriteExt: '' },
      ".ts": { rewriteExt: '' },
      ".d.ts": { rewriteExt: '' },
      ...options?.exts,
    }
  }

  resolveExpression(pathLike: string): URLTransformer | null {
    for (const [ext, { rewriteExt }] of Object.entries(this.exts)) {
      const location = new URL(`${pathLike}${ext}`, this.baseLocation)
      // const relativePath = new URL(`${pathLike}${rewriteExt}`, this.baseLocation)
      logger.log(`Try find file path ${location}`)
      if (existsSync(location)) {
        logger.log(`Found file path ${location}`)
        const locationSource = new URL(`${pathLike}${rewriteExt}`, this.baseLocation)
        return new URLTransformer(location, locationSource)
      }
    }
    return null
  }
}
