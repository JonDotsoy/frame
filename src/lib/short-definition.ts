import { URLTransformer } from "./url-transformer"

export namespace ShortDefinition {
  export type Parameter = {
    module: URLTransformer
    exportName: string
  }

  export type Class = {
    parameters: Parameter[]
  }

  export type Module = {
    classes: Record<string, Class>
  }
}
