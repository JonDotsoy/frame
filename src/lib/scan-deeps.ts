import * as swc from "@swc/core"
import { parseFile, TsTypeAnnotation } from "@swc/core"
import { dirname } from "path"
import { Cache, cacheDefault } from "./dto/cache"
import { ShortDefinition } from "./short-definition"
import { RequireManager } from "./require-manager"
import { URLTransformer } from "./url-transformer"

export async function scanDeeps(
  fileLocationUrl: URLTransformer,
  cache: Cache = cacheDefault()
): Promise<Cache> {
  const requireManager = new RequireManager(fileLocationUrl.toDirname().toURL())
  const fileLocation = requireManager.resolveExpression(fileLocationUrl.toPath())
  if (!fileLocation) throw new Error(`Cannot found source ${fileLocationUrl}`)
  if (cache.has(fileLocation.toString())) return cache

  const program = await parseFile(fileLocation.toPath(), {
    syntax: "typescript",
    target: "es2022",
  })

  const shortModule: ShortDefinition.Module = {
    classes: {},
  }

  const importIdentifiers = new Map<string, ShortDefinition.Parameter>()

  for (const moduleItem of program.body) {
    if (moduleItem.type === "ImportDeclaration") {
      const validRequireModule = requireManager.resolveExpression(moduleItem.source.value)
      if (validRequireModule) {
        for (const specifier of moduleItem.specifiers) {
          if (specifier.type === "ImportSpecifier") {
            importIdentifiers.set(specifier.local.value, {
              exportName: specifier.imported?.value ?? specifier.local.value,
              module: validRequireModule,
            })
          }
          if (specifier.type === "ImportDefaultSpecifier") {
            importIdentifiers.set(specifier.local.value, {
              exportName: "default",
              module: validRequireModule,
            })
          }
        }
      }
    }
    if (moduleItem.type === "ExportDeclaration") {
      const declaration = moduleItem.declaration

      if (declaration.type === "ClassDeclaration") {
        const shortClassName = declaration.identifier.value
        const shortClass: ShortDefinition.Class = {
          parameters: [],
        }
        const itemConstructor = declaration.body.find(
          (item) => item.type === "Constructor"
        ) as swc.Constructor | undefined
        if (itemConstructor) {
          for (const param of itemConstructor.params) {
            if (param.type === "TsParameterProperty") {
              const tsParamProperty = param.param
              if (tsParamProperty.type === "Identifier") {
                if (tsParamProperty.typeAnnotation) {
                  const typeAnnotation =
                    tsParamProperty.typeAnnotation.typeAnnotation
                  if (typeAnnotation.type == "TsTypeReference") {
                    const tsEntityName = typeAnnotation.typeName
                    if (tsEntityName.type === "Identifier") {
                      const parameter = importIdentifiers.get(
                        tsEntityName.value
                      )
                      if (parameter) {
                        shortClass.parameters.push(parameter)
                      }
                    }
                  }
                }
              }
            }
            else if (param.type === "Parameter") {
              const pat = param.pat
              if (pat.type === "Identifier") {
                console.log(pat)
                const typeAnnotation: TsTypeAnnotation | undefined = Object.getOwnPropertyDescriptor(pat, "typeAnnotation")?.value
                if (typeAnnotation) {
                  const typeReference = typeAnnotation.typeAnnotation
                  if (typeReference.type === "TsTypeReference") {
                    const typeName = typeReference.typeName
                    if (typeName.type==="Identifier") {
                      const parameter = importIdentifiers.get(typeName.value)
                      if (parameter) {
                        shortClass.parameters.push(parameter)
                      }
                    }
                  }
                }
              }
            } else {
              throw new Error("Not parameter identifier")
            }
          }
        }

        shortModule.classes[shortClassName] = shortClass
      }
    }
  }

  cache.set(fileLocation.toString(), shortModule)

  for (const [_, parameter] of importIdentifiers) {
    await scanDeeps(parameter.module, cache)
  }

  return cache
}
