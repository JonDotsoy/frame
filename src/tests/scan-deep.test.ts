import { describe, expect, it } from "vitest";
import { scanDeeps } from "../lib/scan-deeps";
import { pathToFileURL } from "node:url"
import { URLTransformer } from "../lib/url-transformer";
import { inspect } from "node:util";


describe("Make deep file", () => {
  it("should scan the sample cat.class.ts file", async () => {
    const appLocation = pathToFileURL(`${__dirname}/__samples__1/cat.app`);

    const cache = await scanDeeps(new URLTransformer(appLocation))

    expect(cache).matchSnapshot()
  })

  it("should scan the sample cat.class.js with cat.class.d.ts file", async () => {
    const appLocation = pathToFileURL(`${__dirname}/__samples__2/cat.app`);

    const cache = await scanDeeps(new URLTransformer(appLocation))

    expect(cache).matchSnapshot()
  })
})
