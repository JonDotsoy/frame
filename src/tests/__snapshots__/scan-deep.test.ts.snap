// Vitest Snapshot v1

exports[`Make deep file > should scan the sample cat.class.js with cat.class.d.ts file 1`] = `
Map {
  "file:///Users/jondotsoy/Repositories/jondotsoy/frame/src/tests/__samples__2/cat.app.d.ts" => {
    "classes": {
      "CatApp": {
        "parameters": [
          {
            "exportName": "CatController",
            "module": URLTransformer {
              "url": "file:///Users/jondotsoy/Repositories/jondotsoy/frame/src/tests/__samples__2/cat.controller.d.ts",
              "urlSource": "file:///Users/jondotsoy/Repositories/jondotsoy/frame/src/tests/__samples__2/cat.controller",
            },
          },
        ],
      },
    },
  },
  "file:///Users/jondotsoy/Repositories/jondotsoy/frame/src/tests/__samples__2/cat.controller.d.ts" => {
    "classes": {
      "CatController": {
        "parameters": [],
      },
    },
  },
}
`;

exports[`Make deep file > should scan the sample cat.class.ts file 1`] = `
Map {
  "file:///Users/jondotsoy/Repositories/jondotsoy/frame/src/tests/__samples__1/cat.app.ts" => {
    "classes": {
      "CatApp": {
        "parameters": [
          {
            "exportName": "CatController",
            "module": URLTransformer {
              "url": "file:///Users/jondotsoy/Repositories/jondotsoy/frame/src/tests/__samples__1/cat.controller.ts",
              "urlSource": "file:///Users/jondotsoy/Repositories/jondotsoy/frame/src/tests/__samples__1/cat.controller",
            },
          },
        ],
      },
    },
  },
  "file:///Users/jondotsoy/Repositories/jondotsoy/frame/src/tests/__samples__1/cat.controller.ts" => {
    "classes": {
      "CatController": {
        "parameters": [],
      },
    },
  },
}
`;
